Allowlist tutorial (#135)
sei-protocol/sei-docs: pages/dev-tutorials/tokenfactory-allow-list.mdx

import { Callout } from 'nextra/components';

# Token Allowlists

The SEI Token Factory module enables token creators to restrict who can interact with their custom tokens through the
use of allowlists.

This tutorial will walk you through:

1. Creating a new denom with an allowlist.
2. Updating an existing denom to add or modify an allowlist.
3. Querying the allowlist of a denom.
4. Understanding the configuration and limitations of allowlists.

## Prerequisites

To create a token on the devnet, ensure you have the following setup:

- The `seid` CLI installed.
- A wallet with SEI tokens on devnet.
- Access to a running Sei blockchain node.

<Callout type='info'>You can obtain devnet tokens from one of the faucets listed [here](../dev-ecosystem-providers/faucets).</Callout>

## Creating a Denom with Allow List

1. First, create a JSON file that contains the list of addresses you want to include in your allow list.
The format of the allow list is as follows:

> ```json

Reviewer: A nit - as we have numbered these sections, the codeblocks should be indented so they are properly formatted underneath the point.

{
    "addresses": [
        "address1",
        "address2",
        "address3"
    ]
}
```

> Ensure that all addresses are valid SEI (Bech32) or EVM (0x-prefixed) addresses. Note that 0x addresses will be converted to Bech32 addresses when the allow list is persisted.
> The maximum number of addresses allowed in the allow list is 2000 by default. This can be configured by the chain administrator.

> 2. Create a new denom with the specified allow list.

Reviewer: ```suggestion

2. Create a new denom with the specified allow list.
```

> ```bash

Reviewer: ```suggestion

```bash
```

seid tx tokenfactory create-denom $SUBDENOM \
  --allow-list=$ALLOW_LIST_FILE_PATH \
  --from=$CREATOR_ACCOUNT \
  --chain-id=$CHAIN_ID \
  --fees=$FEE_AMOUNT \
  --gas=$GAS_LIMIT \
  -y
```

For example:

```bash
seid tx tokenfactory create-denom mytoken \
  --allow-list=./allow_list.json \
  --from mykey \
  --chain-id sei-chain \
  --fees 10000usei \
  --gas auto \
  -y
```

> ### Understanding Command Line Arguments

Reviewer: ```suggestion

### Understanding Command Line Arguments
```

- `--node https://rpc.arctic-1.seinetwork.io/ `: This points to the RPC URL of the node you are interacting with.
- `--broadcast-mode block`: This determines how your transaction is broadcasted to the network. The `block` mode means the transaction will wait to be included in a block before returning a response. This is a safer option as it confirms your transaction is processed.
- `--fees 20000usei`: This is used to specify the transaction fee.

Understanding these arguments will help you execute the commands more confidently and customize them as needed for different scenarios.

<Callout type='info'>For detailed descriptions of these arguments, use `seid help` in the CLI.</Callout>

## Updating an Existing Denom Allowlist

To update the allow list of a token, you can use the following command:

```bash
seid tx tokenfactory update-denom $DENOM \
  --allow-list=$ALLOW_LIST_FILE_PATH \
  --from=$ACCOUNT \
  --chain-id=$CHAIN_ID \
  --node=$NODE_RPC_URL \
  --broadcast-mode=block \
  --fees=$FEE_AMOUNT \
  --gas=$GAS_LIMIT \
  -y
```

For example:

> ```bash

Reviewer: ```suggestion
```bash

```

seid tx tokenfactory update-denom mytoken \
  --allow-list=./updated_allow_list.json \
  --from mykey \
  --chain-id arctic-1 \
  --node https://rpc.arctic-1.seinetwork.io/ \
  --broadcast-mode=block \
  --fees 20000usei \
  --gas auto \
  -y

> ````

Reviewer: ```suggestion
```
```

> To re-enable all addresses to transfer the token, you can simply submit an empty allowlist.

Reviewer: ```suggestion
To re-enable all addresses to transfer the token, you can simply submit an empty allowlist.

```

import { Callout } from 'nextra/components';

# Token Allowlists

The SEI Token Factory module enables token creators to restrict who can interact with their custom tokens through the
use of allowlists.

This tutorial will walk you through:

1. Creating a new denom with an allowlist.
2. Updating an existing denom to add or modify an allowlist.
3. Querying the allowlist of a denom.
4. Understanding the configuration and limitations of allowlists.

## Prerequisites

To create a token on the devnet, ensure you have the following setup:

- The `seid` CLI installed.
- A wallet with SEI tokens on devnet.
- Access to a running Sei blockchain node.

<Callout type='info'>You can obtain devnet tokens from one of the faucets listed [here](../dev-ecosystem-providers/faucets).</Callout>

## Creating a Denom with Allow List

1. First, create a JSON file that contains the list of addresses you want to include in your allow list.
The format of the allow list is as follows:

> 	```json

Reviewer: ```suggestion
    ```json
```

Reviewer: I think spacing should be 4 width instead of 8 (as its too indented)
<img width="1035" alt="image" src="https://github.com/user-attachments/assets/3a74b316-9802-48ef-a496-bc769b437659">

import { Callout } from 'nextra/components';

> # Introducing Token Add-ons

Reviewer: ```suggestion
# Token Allowlists
```

Reviewer: While it's true that this tutorial is technically introducing token add-ons, since the primary focus is on allowlists, it would make sense to reflect that more directly in the title.

Reviewer: This could eventually be part of an add-ons section in the docs.

Reviewer: Also see [this](https://github.com/sei-protocol/sei-docs/pull/135/files#r1759307446) for suggesting denom.

> As you have learned form the [Token Factory tutorial](tokenfactory-tutorial.mdx), the `tokenfactory` module enables any account to create new tokens with a unique identifier based on the creator's address. 

Reviewer: On nomenclature -

- I believe we can use "allowlist" instead of "allow list", similar to "whitelist" (compound noun).
- I believe denom is more accurate instead of token, but token is probably better for things like SEO.

Introducing Token Add-ons, a set of features that can be added to your token to enhance its functionality.
With the first add-on being an allow list, which restricts the transferability of the token to a specific set of addresses.

In this tutorial, we will learn how to create a token with an allow list.

## Requirements

To create a token on the devnet, ensure you have the following setup:

- The `seid` CLI
- A wallet with SEI tokens on devnet

<Callout type='info'>You can obtain devnet tokens from one of the faucets listed [here](../dev-ecosystem-providers/faucets).</Callout>

> 

Reviewer: Can link example of node RPC to use and can also call out running locally.

## Creating a Denom with Allow List

First, we need to prepare a list of addresses that are allowed to transfer the token and save it to a JSON file.
The format of the allow list is as follows:

```json
{
    "addresses": [
        "address1",
        "address2",
        "address3"
    ]
}
```

Where each address is a valid Sei bech32 address or EVM 0x prefixed address. Please note, that 0x addresses will be converted to bech32 addresses when the allow list is persisted.

```bash copy
seid tx tokenfactory create-denom $DENOM --from=$ACCOUNT --allow-list=$ALLOW_LIST_FILE_PATH --chain-id=arctic-1 --node=https://rpc.arctic-1.seinetwork.io/ --broadcast-mode=block --fees=20000usei
```

> This command creates a new coin with specific allow list.

Reviewer: The example commands suggested in the sections may be unnecessary and could clutter the content. We could also rely on the main code block and incorporate the example.

### Understanding Command Line Arguments

> 

Reviewer: In addition, this section could be restructured into "Understanding Allowlist Validations & Limitations," 

Summarizing key aspects and providing a reference point for other sections. For example, when creating the JSON file of addresses, [this](https://github.com/sei-protocol/sei-docs/pull/135/files#diff-e71c3e3d8503032b25c3bfb1219b087164713ed2fca9c58089144e6f671f60c3R36) could be replaced with: "See the 'Understanding Allowlist Validations & Limitations' section for details." which captures the associated details.

```
- **Valid Addresses**: All addresses in the allowlist must be valid in their respective formats (SEI Bech32 or EVM 0x prefixed). Invalid addresses will result in a validation error.
- **Address Conversion**: EVM addresses (0x prefixed) are automatically converted to SEI Bech32 addresses when the allow list is processed.
- **No Duplicates**: Duplicate addresses in the allowlist are not allowed and will result in a validation error.
- **Maximum Size**: By default, the maximum number of addresses allowed in the allowlist is 2,000. This limit can be configured by the chain administrator by setting `DenomAllowListMaxSize`.
- **Non-Empty List**: The allow list cannot be empty when provided. If you wish to remove the allow list, you need to pass an empty JSON object {}.
- And so on...

```

When executing commands in this tutorial, you'll encounter several arguments. Here's a brief overview of what each means:

- `--allow-list $ALLOW_LIST_FILE_PATH`: This specifies the path to the allow list file that we have created in the previous step. E.g. `allowlist.json`.
- `--chain-id arctic-1`: This specifies the network where the command will be executed. In this case, `arctic-1` is the identifier for the Sei devnet.
- `--node https://rpc.arctic-1.seinetwork.io/ `: This points to the RPC URL of the node you are interacting with.
- `--broadcast-mode block`: This determines how your transaction is broadcasted to the network. The `block` mode means the transaction will wait to be included in a block before returning a response. This is a safer option as it confirms your transaction is processed.
- `--fees 20000usei`: This is used to specify the transaction fee.

Understanding these arguments will help you execute the commands more confidently and customize them as needed for different scenarios.

<Callout type='info'>For detailed descriptions of these arguments, use `seid help` in the CLI.</Callout>

## Updating Token Allow List

> To update the allow list of a token, you can use the following command:

Reviewer: ```suggestion
1. To update the allowlist of a token, you can use the following command:
```

> ```bash copy

Reviewer: ```suggestion
```bash
```

> seid tx tokenfactory update-denom $DENOM --from=$ACCOUNT --allow-list=$ALLOW_LIST_FILE_PATH --chain-id=arctic-1 --node=https://rpc.arctic-1.seinetwork.io/ --broadcast-mode=block --fees=20000usei

Reviewer: ```suggestion
seid tx tokenfactory update-denom $DENOM \
  --allow-list=$ALLOW_LIST_FILE_PATH \
  --from=$ACCOUNT \
  --chain-id=$CHAIN_ID \
  --node=$NODE_RPC_URL \
  --broadcast-mode=block \
  --fees=$FEE_AMOUNT \
  --gas=$GAS_LIMIT \
  -y
```

```

> This command updates the allow list of a token.

Reviewer: ```suggestion
For example:
```bash
seid tx tokenfactory update-denom mytoken \
  --allow-list=./updated_allow_list.json \
  --from mykey \
  --chain-id arctic-1 \
  --node https://rpc.arctic-1.seinetwork.io/ \
  --broadcast-mode=block \
  --fees 20000usei \
  --gas auto \
  -y
```

This command updates the allowlist of 'mytoken'.
```

> To allow all addresses agian to transfer the token, you can simply submit an empty allow list.

Reviewer: ```suggestion
To re-enable all addresses to transfer the token, you can simply submit an empty allowlist.
```

## Querying Token Allow List

> You may query the allow list of a token using the node REST endpoint. E.g.:

Reviewer: ```suggestion
You may query the allowlist of a token using the node REST endpoint. E.g.:
```

> ```bash copy

Reviewer: ```suggestion

```bash
```

curl -X 'GET' \
  'https://rest-arctic-1.sei-apis.com/sei-protocol/seichain/tokenfactory/denoms/allow_list?denom=factory/{ACCOUNT}/{DENOM}' \
  -H 'accept: application/json'
```

> 

Reviewer: ```suggestion
```
